From 4014c7f00f03deeba45fba7d02f6b3930e718b9e Mon Sep 17 00:00:00 2001
From: Huaxiang Sun <hsun@cloudera.com>
Date: Thu, 9 Mar 2017 17:46:55 -0800
Subject: [PATCH 281/303] Revert "HBASE-16266 Do not throw
 ScannerTimeoutException when catch
 UnknownScannerException"

Revert because it breaks compatibility with old cdh releases without scanner heartbeat support.

This reverts commit fa4b9bb2bb6a623e95b6b08941c45b4979c96b15.
---
 .../apache/hadoop/hbase/client/ClientScanner.java  |   38 +++--
 .../hbase/TestPartialResultsFromClientSide.java    |  147 --------------------
 .../hadoop/hbase/client/TestScannerTimeout.java    |   33 +++++
 3 files changed, 59 insertions(+), 159 deletions(-)
 delete mode 100644 hbase-server/src/test/java/org/apache/hadoop/hbase/TestPartialResultsFromClientSide.java

diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ClientScanner.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ClientScanner.java
index e4eaf46..ce04409 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ClientScanner.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ClientScanner.java
@@ -418,20 +418,34 @@ public class ClientScanner extends AbstractClientScanner {
         clearPartialResults();
         // DNRIOEs are thrown to make us break out of retries. Some types of DNRIOEs want us
         // to reset the scanner and come back in again.
-
-        // If exception is any but the list below throw it back to the client; else setup
-        // the scanner and retry.
-        Throwable cause = e.getCause();
-        if ((cause != null && cause instanceof NotServingRegionException) ||
-            (cause != null && cause instanceof RegionServerStoppedException) ||
-            e instanceof OutOfOrderScannerNextException ||
-            e instanceof UnknownScannerException ) {
-          // Pass. It is easier writing the if loop test as list of what is allowed rather than
-          // as a list of what is not allowed... so if in here, it means we do not throw.
+        if (e instanceof UnknownScannerException) {
+          long timeout = lastNext + scannerTimeout;
+          // If we are over the timeout, throw this exception to the client wrapped in
+          // a ScannerTimeoutException. Else, it's because the region moved and we used the old
+          // id against the new region server; reset the scanner.
+          if (timeout < System.currentTimeMillis()) {
+            LOG.info("For hints related to the following exception, please try taking a look at: "
+                    + "https://hbase.apache.org/book.html#trouble.client.scantimeout");
+            long elapsed = System.currentTimeMillis() - lastNext;
+            ScannerTimeoutException ex =
+                new ScannerTimeoutException(elapsed + "ms passed since the last invocation, "
+                    + "timeout is currently set to " + scannerTimeout);
+            ex.initCause(e);
+            throw ex;
+          }
         } else {
-          throw e;
+          // If exception is any but the list below throw it back to the client; else setup
+          // the scanner and retry.
+          Throwable cause = e.getCause();
+          if ((cause != null && cause instanceof NotServingRegionException) ||
+              (cause != null && cause instanceof RegionServerStoppedException) ||
+              e instanceof OutOfOrderScannerNextException) {
+            // Pass. It is easier writing the if loop test as list of what is allowed rather than
+            // as a list of what is not allowed... so if in here, it means we do not throw.
+          } else {
+            throw e;
+          }
         }
-
         // Else, its signal from depths of ScannerCallable that we need to reset the scanner.
         if (this.lastResult != null) {
           // The region has moved. We need to open a brand new scanner at the new location.
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/TestPartialResultsFromClientSide.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/TestPartialResultsFromClientSide.java
deleted file mode 100644
index 6447cae..0000000
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/TestPartialResultsFromClientSide.java
+++ /dev/null
@@ -1,147 +0,0 @@
-/**
- * Licensed to the Apache Software Foundation (ASF) under one
- * or more contributor license agreements.  See the NOTICE file
- * distributed with this work for additional information
- * regarding copyright ownership.  The ASF licenses this file
- * to you under the Apache License, Version 2.0 (the
- * "License"); you may not use this file except in compliance
- * with the License.  You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package org.apache.hadoop.hbase;
-
-import static org.junit.Assert.assertEquals;
-
-import java.io.IOException;
-import java.util.ArrayList;
-import java.util.List;
-import org.apache.hadoop.hbase.client.Put;
-import org.apache.hadoop.hbase.client.Result;
-import org.apache.hadoop.hbase.client.ResultScanner;
-import org.apache.hadoop.hbase.client.Scan;
-import org.apache.hadoop.hbase.client.Table;
-import org.apache.hadoop.hbase.testclassification.MediumTests;
-import org.apache.hadoop.hbase.util.Bytes;
-import org.junit.AfterClass;
-import org.junit.BeforeClass;
-import org.junit.Test;
-import org.junit.experimental.categories.Category;
-
-/**
- * These tests are focused on testing how partial results appear to a client. Partial results are
- * {@link Result}s that contain only a portion of a row's complete list of cells. Partial results
- * are formed when the server breaches its maximum result size when trying to service a client's RPC
- * request. It is the responsibility of the scanner on the client side to recognize when partial
- * results have been returned and to take action to form the complete results.
- * <p>
- * Unless the flag {@link Scan#setAllowPartialResults(boolean)} has been set to true, the caller of
- * {@link ResultScanner#next()} should never see partial results.
- */
-@Category(MediumTests.class)
-public class TestPartialResultsFromClientSide {
-
-  private final static HBaseTestingUtility TEST_UTIL = new HBaseTestingUtility();
-  private final static int MINICLUSTER_SIZE = 5;
-  private static Table TABLE = null;
-
-  /**
-   * Table configuration
-   */
-  private static TableName TABLE_NAME = TableName.valueOf("testTable");
-
-  private static int NUM_ROWS = 5;
-  private static byte[] ROW = Bytes.toBytes("testRow");
-  private static byte[][] ROWS = HTestConst.makeNAscii(ROW, NUM_ROWS);
-
-  // Should keep this value below 10 to keep generation of expected kv's simple. If above 10 then
-  // table/row/cf1/... will be followed by table/row/cf10/... instead of table/row/cf2/... which
-  // breaks the simple generation of expected kv's
-  private static int NUM_FAMILIES = 10;
-  private static byte[] FAMILY = Bytes.toBytes("testFamily");
-  private static byte[][] FAMILIES = HTestConst.makeNAscii(FAMILY, NUM_FAMILIES);
-
-  private static int NUM_QUALIFIERS = 10;
-  private static byte[] QUALIFIER = Bytes.toBytes("testQualifier");
-  private static byte[][] QUALIFIERS = HTestConst.makeNAscii(QUALIFIER, NUM_QUALIFIERS);
-
-  private static int VALUE_SIZE = 1024;
-  private static byte[] VALUE = Bytes.createMaxByteArray(VALUE_SIZE);
-
-  private static long timeout = 2000;
-
-  @BeforeClass
-  public static void setUpBeforeClass() throws Exception {
-    TEST_UTIL.getConfiguration().setLong(HConstants.HBASE_CLIENT_SCANNER_TIMEOUT_PERIOD, timeout);
-    TEST_UTIL.startMiniCluster(MINICLUSTER_SIZE);
-    TEST_UTIL.getHBaseAdmin().setBalancerRunning(false, true);
-    TABLE = createTestTable(TABLE_NAME, ROWS, FAMILIES, QUALIFIERS, VALUE);
-  }
-
-  static Table createTestTable(TableName name, byte[][] rows, byte[][] families,
-      byte[][] qualifiers, byte[] cellValue) throws IOException {
-    Table ht = TEST_UTIL.createTable(name, families);
-    List<Put> puts = createPuts(rows, families, qualifiers, cellValue);
-    ht.put(puts);
-
-    return ht;
-  }
-
-  @AfterClass
-  public static void tearDownAfterClass() throws Exception {
-    TEST_UTIL.shutdownMiniCluster();
-  }
-
-  /**
-   * Make puts to put the input value into each combination of row, family, and qualifier
-   * @param rows
-   * @param families
-   * @param qualifiers
-   * @param value
-   * @return
-   * @throws IOException
-   */
-  static ArrayList<Put> createPuts(byte[][] rows, byte[][] families, byte[][] qualifiers,
-      byte[] value) throws IOException {
-    Put put;
-    ArrayList<Put> puts = new ArrayList<>();
-
-    for (int row = 0; row < rows.length; row++) {
-      put = new Put(rows[row]);
-      for (int fam = 0; fam < families.length; fam++) {
-        for (int qual = 0; qual < qualifiers.length; qual++) {
-          KeyValue kv = new KeyValue(rows[row], families[fam], qualifiers[qual], qual, value);
-          put.add(kv);
-        }
-      }
-      puts.add(put);
-    }
-
-    return puts;
-  }
-
-  @Test
-  public void testDontThrowUnknowScannerExceptionToClient() throws Exception {
-    Table table =
-        createTestTable(TableName.valueOf("testDontThrowUnknowScannerException"), ROWS, FAMILIES,
-            QUALIFIERS, VALUE);
-    Scan scan = new Scan();
-    scan.setCaching(1);
-    ResultScanner scanner = table.getScanner(scan);
-    scanner.next();
-    Thread.sleep(timeout * 3);
-    int count = 1;
-    while (scanner.next() != null) {
-      count++;
-    }
-    assertEquals(NUM_ROWS, count);
-    scanner.close();
-  }
-
-}
\ No newline at end of file
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestScannerTimeout.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestScannerTimeout.java
index 0054d09..286bfd6 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestScannerTimeout.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestScannerTimeout.java
@@ -94,6 +94,39 @@ public class TestScannerTimeout {
   }
 
   /**
+   * Test that we do get a ScannerTimeoutException
+   * @throws Exception
+   */
+  @Test(timeout=300000)
+  public void test2481() throws Exception {
+    LOG.info("START ************ test2481");
+    Scan scan = new Scan();
+    scan.setCaching(1);
+    Table table =
+      new HTable(new Configuration(TEST_UTIL.getConfiguration()), TABLE_NAME);
+    ResultScanner r = table.getScanner(scan);
+    int count = 0;
+    try {
+      Result res = r.next();
+      while (res != null) {
+        count++;
+        if (count == 5) {
+          // Sleep just a bit more to be sure
+          Thread.sleep(SCANNER_TIMEOUT + THREAD_WAKE_FREQUENCY + 100);
+        }
+        res = r.next();
+      }
+    } catch (ScannerTimeoutException e) {
+      LOG.info("Got the timeout " + e.getMessage(), e);
+      return;
+    }  finally {
+      table.close();
+    }
+    fail("We should be timing out");
+    LOG.info("END ************ test2481");
+  }
+
+  /**
    * Test that scanner can continue even if the region server it was reading
    * from failed. Before 2772, it reused the same scanner id.
    * @throws Exception
-- 
1.7.9.5

